File: internal/workspace/workspace.go
// Package workspace assembles and dismantles a workspace: N git worktrees plus
// the untracked context they need to run.
package workspace

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/milindmadhukar/paseo.nvim/internal/gitx"
	"github.com/milindmadhukar/paseo.nvim/internal/manifest"
	"github.com/milindmadhukar/paseo.nvim/internal/registry"
)

// CreateOptions configures assembly.
type CreateOptions struct {
	Name  string
	Root  string   // project root
	Only  []string // exactly these repos
	With  []string // these in addition to the defaults
	Setup bool     // run each repo's setup commands
}

// Create assembles a workspace.
func Create(ctx context.Context, m *manifest.Manifest, opts CreateOptions) (*registry.Workspace, error) {
	if opts.Name == "" {
		return nil, fmt.Errorf("a workspace needs a name")
	}
	if strings.ContainsAny(opts.Name, "/\\ ") {
		return nil, fmt.Errorf("workspace names may not contain slashes or spaces: %q", opts.Name)
	}

	names, err := m.Select(opts.Only, opts.With)
	if err != nil {
		return nil, err
	}
	if len(names) == 0 {
		return nil, fmt.Errorf("no repos selected; every repo is default = false, so name some with --repos")
	}

	root := filepath.Join(opts.Root, m.WorkspacesDir, opts.Name)
	if _, err := os.Stat(root); err == nil {
		return nil, fmt.Errorf("%s already exists", root)
	}
	if err := os.MkdirAll(root, 0o755); err != nil {
		return nil, err
	}

	branch := m.BranchPrefix + opts.Name
	ws := registry.Workspace{
		Name:      opts.Name,
		Project:   opts.Root,
		Root:      root,
		Branch:    branch,
		CreatedAt: time.Now().UTC(),
	}

	// Single-repo projects put the worktree AT the workspace root rather than
	// in a subdirectory, so paths stay `app/main.py` instead of
	// `myrepo/app/main.py` for no reason. paseo.nvim's repo detection handles
	// both shapes.
	single := len(names) == 1 && len(m.Repos) == 1

	for _, name := range names {
		repo := m.Repos[name]
		origin := filepath.Join(opts.Root, name)
		dest := root
		if !single {
			dest = filepath.Join(root, name)
		}

		if err := addWorktree(ctx, origin, dest, branch, repo.Base); err != nil {
			// Leave what was assembled: dismantling a partial workspace on
			// failure would throw away the members that DID come up, and
			// `ws rm` can do it deliberately.
			return nil, fmt.Errorf("%s: %w", name, err)
		}

		if repo.Submodules {
			if _, err := gitx.Run(ctx, dest, "submodule", "update", "--init", "--recursive"); err != nil {
				return nil, fmt.Errorf("%s: submodules: %w", name, err)
			}
		}
		if err := materialise(origin, dest, repo); err != nil {
			return nil, fmt.Errorf("%s: %w", name, err)
		}
		if err := excludeManaged(ctx, dest, repo); err != nil {
			return nil, fmt.Errorf("%s: %w", name, err)
		}
		if opts.Setup {
			if err := runSetup(ctx, dest, repo.Setup); err != nil {
				return nil, fmt.Errorf("%s: setup: %w", name, err)
			}
		}

		ws.Repos = append(ws.Repos, registry.Repo{
			Name:   name,
			Path:   dest,
			Origin: origin,
			Branch: branch,
			Base:   repo.Base,
			State:  registry.StateActive,
		})
	}

	// Members named in the manifest but not selected are recorded as declared,
	// so `ws ls` can show what a workspace could still pull in.
	for _, name := range m.Names() {
		if !contains(names, name) {
			ws.Repos = append(ws.Repos, registry.Repo{
				Name:   name,
				Origin: filepath.Join(opts.Root, name),
				State:  registry.StateDeclared,
			})
		}
	}

	if err := linkShared(opts.Root, root, m.Shared); err != nil {
		return nil, err
	}

	return &ws, nil
}

// addWorktree creates the worktree and its branch.
//
// The base is resolved to `origin/<base>` when that ref exists. Branching from
// the LOCAL branch would base the work on whatever happens to be checked out in
// the primary tree, including unpushed commits and a stale position -- and the
// primary checkout being behind is the normal case, not the exception.
func addWorktree(ctx context.Context, origin, dest, branch, base string) error {
	if !gitx.IsRepo(ctx, origin) {
		return fmt.Errorf("%s is not a git work tree", origin)
	}

	start := base
	if _, err := gitx.Run(ctx, origin, "rev-parse", "--verify", "--quiet", "refs/remotes/origin/"+base); err == nil {
		start = "origin/" + base
	}

	if _, err := gitx.Run(ctx, origin, "worktree", "add", "-b", branch, dest, start); err != nil {
		// An existing branch is the common second run; reuse it rather than
		// failing, which is what -b cannot do.
		if strings.Contains(err.Error(), "already exists") {
			_, retry := gitx.Run(ctx, origin, "worktree", "add", dest, branch)
			return retry
		}
		return err
	}
	return nil
}

// materialise copies, links and clones the untracked context a repo needs.
func materialise(origin, dest string, repo manifest.Repo) error {
	for _, name := range repo.Copy {
		src, dst := filepath.Join(origin, name), filepath.Join(dest, name)
		if _, err := os.Lstat(src); err != nil {
			continue // nothing to copy is not an error
		}
		// COPIED, never linked. An agent editing PORT= in a shared .env breaks
		// the primary checkout and every sibling workspace at once.
		if err := copyFile(src, dst); err != nil {
			return fmt.Errorf("copy %s: %w", name, err)
		}
	}

	for _, name := range repo.Link {
		src, dst := filepath.Join(origin, name), filepath.Join(dest, name)
		if _, err := os.Stat(src); err != nil {
			continue
		}
		if err := replaceSymlink(src, dst); err != nil {
			return fmt.Errorf("link %s: %w", name, err)
		}
	}

	for _, name := range repo.CloneSymlinks {
		src := filepath.Join(origin, name)
		target, err := os.Readlink(src)
		if err != nil {
			continue
		}
		// The link's TARGET, recreated. Linking to the link resolves correctly
		// right up until the primary checkout moves or is removed.
		if err := replaceSymlink(target, filepath.Join(dest, name)); err != nil {
			return fmt.Errorf("clone symlink %s: %w", name, err)
		}
	}

	return nil
}

// excludeBegin and excludeEnd delimit ws's block in a repo's info/exclude.
const (
	excludeBegin = "# >>> ws managed >>>"
	excludeEnd   = "# <<< ws managed <<<"
)

// excludeManaged stops ws's own symlinks showing up as untracked.
//
// A repo that gitignores `node_modules/` -- with the trailing slash, which is
// how everyone writes it -- does NOT ignore a SYMLINK of that name, because the
// slash means "directory". Every linked directory therefore shows as
// `?? node_modules` in the worktree forever, and `ws rm` refuses to remove a
// workspace over files it created itself.
//
// This has to go in the COMMON info/exclude: a per-worktree $GIT_DIR/info/exclude
// is silently ignored -- git reads that file only from the common dir. Verified
// rather than assumed. The cost in the primary checkout is nil: the repo's own
// .gitignore already covers the directory, and this only adds the slash-less
// form. The block is delimited so it is obviously ws's and can be deleted.
func excludeManaged(ctx context.Context, dest string, repo manifest.Repo) error {
	managed := append(append([]string{}, repo.Link...), repo.CloneSymlinks...)
	if len(managed) == 0 {
		return nil
	}

	common, err := gitx.Run(ctx, dest, "rev-parse", "--path-format=absolute", "--git-common-dir")
	if err != nil {
		return err
	}
	path := filepath.Join(common, "info", "exclude")
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}

	existing, _ := os.ReadFile(path)
	var kept []string
	inBlock := false
	for _, line := range strings.Split(string(existing), "\n") {
		switch {
		case line == excludeBegin:
			inBlock = true
		case line == excludeEnd:
			inBlock = false
		case !inBlock:
			kept = append(kept, line)
		}
	}
	for len(kept) > 0 && strings.TrimSpace(kept[len(kept)-1]) == "" {
		kept = kept[:len(kept)-1]
	}

	block := append([]string{excludeBegin, "# symlinks ws creates in worktrees; delete this block to undo."}, managed...)
	block = append(block, excludeEnd, "")
	return os.WriteFile(path, []byte(strings.Join(append(kept, block...), "\n")), 0o644)
}

// linkShared symlinks project-level siblings into the workspace root.
func linkShared(project, root string, shared []string) error {
	for _, name := range shared {
		src := filepath.Join(project, name)
		if _, err := os.Stat(src); err != nil {
			continue
		}
		if err := replaceSymlink(src, filepath.Join(root, name)); err != nil {
			return fmt.Errorf("shared %s: %w", name, err)
		}
	}
	return nil
}

func runSetup(ctx context.Context, dir string, commands []string) error {
	for _, command := range commands {
		cmd := exec.CommandContext(ctx, "sh", "-c", command)
		cmd.Dir = dir
		cmd.Stdout, cmd.Stderr = os.Stdout, os.Stderr
		if err := cmd.Run(); err != nil {
			return fmt.Errorf("%q: %w", command, err)
		}
	}
	return nil
}

// RemoveOptions configures dismantling.
type RemoveOptions struct {
	Force bool
}

// Remove dismantles a workspace.
//
// Never `rm -rf`. `git worktree remove` is what keeps .git/worktrees consistent
// -- deleting the directory instead is exactly how the two stale prunable
// entries already on disk got there.
func Remove(ctx context.Context, ws *registry.Workspace, opts RemoveOptions) error {
	if !opts.Force {
		if blockers := unsaved(ctx, ws); len(blockers) > 0 {
			return fmt.Errorf("refusing to remove: %s\npass --force to discard", strings.Join(blockers, "; "))
		}
	}

	for _, repo := range ws.Repos {
		if repo.State != registry.StateActive || repo.Path == "" {
			continue
		}
		args := []string{"worktree", "remove"}
		if opts.Force {
			args = append(args, "--force")
		}
		if _, err := gitx.Run(ctx, repo.Origin, append(args, repo.Path)...); err != nil {
			return fmt.Errorf("%s: %w", repo.Name, err)
		}
	}

	// What is left is ours: the workspace directory and the shared symlinks in
	// it. Removing a symlink never touches what it points at.
	if err := os.RemoveAll(ws.Root); err != nil {
		return err
	}

	// Stale administrative entries are the failure this whole function exists
	// to avoid, so prune regardless.
	for _, repo := range ws.Repos {
		if repo.Origin != "" {
			_, _ = gitx.Run(ctx, repo.Origin, "worktree", "prune")
		}
	}
	return nil
}

// unsaved reports work that removing the workspace would destroy.
func unsaved(ctx context.Context, ws *registry.Workspace) []string {
	var blockers []string
	for _, repo := range ws.Repos {
		if repo.State != registry.StateActive || repo.Path == "" {
			continue
		}
		if out, err := gitx.Run(ctx, repo.Path, "status", "--porcelain"); err == nil && out != "" {
			blockers = append(blockers, fmt.Sprintf("%s has uncommitted changes", repo.Name))
		}

		// Work CREATED HERE and nowhere else: reachable from HEAD, not from any
		// remote, and not from the base this branch was cut from.
		//
		// The base term is what makes this correct. `HEAD --not --remotes`
		// alone lists the ENTIRE history in a repo with no remote configured,
		// so every workspace in such a repo was unremovable -- including one
		// with no commits of its own at all.
		args := []string{"log", "--oneline", "HEAD", "--not", "--remotes"}
		if repo.Base != "" {
			args = append(args, repo.Base)
		}
		if out, err := gitx.Run(ctx, repo.Path, args...); err == nil && out != "" {
			count := len(strings.Split(out, "\n"))
			blockers = append(blockers, fmt.Sprintf("%s has %d unpushed commit(s)", repo.Name, count))
		}
	}
	return blockers
}

func copyFile(src, dst string) error {
	data, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	info, err := os.Stat(src)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
		return err
	}
	return os.WriteFile(dst, data, info.Mode().Perm())
}

// replaceSymlink points dst at target, removing whatever was there.
func replaceSymlink(target, dst string) error {
	if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
		return err
	}
	if _, err := os.Lstat(dst); err == nil {
		if err := os.Remove(dst); err != nil {
			return err
		}
	}
	return os.Symlink(target, dst)
}

func contains(list []string, want string) bool {
	for _, item := range list {
		if item == want {
			return true
		}
	}
	return false
}

File: tests/spec/workspace.lua
--- Workspaces: the model.
---
--- What a workspace is made of, and what `:Paseo ws init` writes down.

local t = require "tests.spec.helpers"
local eq, truthy = t.eq, t.truthy

local root = t.fixtures
local in_dir = t.in_dir

local function test_workspace()
  local workspace = require "paseo.workspace"
  local manifest = workspace.manifest

  -- The TOML subset, including the shapes init emits.
  local parsed = manifest.parse [[
shared = ["Docs"]
[repos.clm_api]
base = "dev"
copy = [".env", ".env.prod"]
submodules = true
[repos.hipa-v2]
base = "dev"
default = false
]]
  eq("manifest: parses repos", manifest.names(parsed), { "clm_api", "hipa-v2" })
  eq("manifest: parses arrays", parsed.repos.clm_api.copy, { ".env", ".env.prod" })
  eq("manifest: parses bools", parsed.repos.clm_api.submodules, true)
  -- Absent means included; only an explicit false opts out.
  eq("manifest: default set excludes `default = false`", manifest.select(parsed), { "clm_api" })
  eq(
    "manifest: --with opts one back in",
    manifest.select(parsed, nil, { "hipa-v2" }),
    { "clm_api", "hipa-v2" }
  )
  eq(
    "manifest: --repos wins over --with",
    manifest.select(parsed, { "clm_api" }, { "hipa-v2" }),
    { "clm_api" }
  )
  truthy(
    "manifest: an unknown repo is an error",
    select(2, manifest.select(parsed, { "nope" })) ~= nil
  )
  eq("manifest: render round-trips", manifest.parse(manifest.render(parsed, {})), parsed)

  -- Assembly, against the multi-repo fixture.
  local project = root .. "/multi"
  local m, notes = workspace.discover(project)
  truthy(
    "workspace: discovers the members",
    m and #manifest.names(m) == 2,
    m and vim.inspect(manifest.names(m))
  )
  truthy("workspace: Docs is shared", m and vim.tbl_contains(m.shared, "Docs"))
  -- `Docs` is a plain directory; `.workspaces` holds the worktrees and must
  -- never be offered as a shared sibling.
  truthy(
    "workspace: .workspaces is not shared",
    m and not vim.tbl_contains(m.shared, ".workspaces")
  )

  -- The fixture's node_modules is too small to be classified heavy, so declare
  -- the link explicitly -- link handling is what is under test.
  for _, name in ipairs(manifest.names(m)) do
    m.repos[name].link = { "node_modules" }
    m.repos[name].copy = { ".env" }
  end
  for _, name in ipairs(manifest.names(m)) do
    local fd = io.open(vim.fs.joinpath(project, name, ".env"), "w")
    if fd then
      fd:write "PORT=3000\n"
      fd:close()
    end
    vim.fn.mkdir(vim.fs.joinpath(project, name, "node_modules"), "p")
  end

  -- The fixture already holds an `otp` workspace, so worktree counts are
  -- compared BEFORE and AFTER rather than against an assumed 1.
  local function worktrees(name)
    local out = vim
      .system({ "git", "-C", vim.fs.joinpath(project, name), "worktree", "list" }, { text = true })
      :wait()
    return #vim.split(vim.trim(out.stdout or ""), "\n")
  end
  local before = { clm = worktrees "clm", clm_api = worktrees "clm_api" }

  local ws, err = workspace.create(m, { name = "spec", root = project })
  truthy("workspace: create succeeds", ws ~= nil, err)
  if not ws then
    return
  end

  for _, name in ipairs { "clm", "clm_api" } do
    local dir = vim.fs.joinpath(ws.root, name)
    local env = vim.uv.fs_lstat(vim.fs.joinpath(dir, ".env"))
    eq("workspace: " .. name .. "/.env is a copy, not a link", env and env.type, "file")
    local link = vim.uv.fs_lstat(vim.fs.joinpath(dir, "node_modules"))
    eq("workspace: " .. name .. "/node_modules is a link", link and link.type, "link")
  end

  local docs = vim.uv.fs_lstat(vim.fs.joinpath(ws.root, "Docs"))
  eq("workspace: Docs is symlinked into the workspace root", docs and docs.type, "link")

  -- A repo that ignores `node_modules/` does NOT ignore a SYMLINK of that name,
  -- so without the exclude block every linked dir shows as untracked forever --
  -- and removal then refuses over files we created ourselves.
  for _, name in ipairs { "clm", "clm_api" } do
    local dirty = vim
      .system({ "git", "-C", vim.fs.joinpath(ws.root, name), "status", "--porcelain" }, { text = true })
      :wait()
    eq("workspace: " .. name .. " worktree is clean", vim.trim(dirty.stdout or ""), "")
  end

  -- The primary checkouts are untouched: the whole point of isolation.
  for _, name in ipairs { "clm", "clm_api" } do
    local branch = vim
      .system({ "git", "-C", vim.fs.joinpath(project, name), "branch", "--show-current" }, { text = true })
      :wait()
    eq(
      "workspace: primary " .. name .. " is still on its own branch",
      vim.trim(branch.stdout or ""),
      "main"
    )
  end

  -- An untouched workspace removes cleanly. The regression: `HEAD --not
  -- --remotes` in a repo with NO remote lists the entire history, so every
  -- workspace in such a repo was unremovable.
  eq("workspace: nothing unsaved in a fresh workspace", workspace.unsaved(ws), {})
  local removed, rm_err = workspace.remove(ws)
  truthy("workspace: remove succeeds", removed, rm_err)

  for _, name in ipairs { "clm", "clm_api" } do
    -- `git worktree remove`, never rm -rf: deleting the directory is how stale
    -- .git/worktrees entries get left behind, and they only surface later as
    -- unrelated-looking failures. Compared against the count from BEFORE
    -- create: the fixture already holds a workspace of its own.
    eq(
      "workspace: " .. name .. " is back to its pre-create worktree count",
      worktrees(name),
      before[name]
    )

    -- And the branch with it. `git worktree remove` leaves it standing, so a
    -- project collected one dead `ws/<name>` per member per workspace anyone
    -- ever made -- kora still had a full set three days after the workspace
    -- that made them was gone.
    local left = vim
      .system(
        { "git", "-C", vim.fs.joinpath(project, name), "branch", "--list", "ws/spec" },
        { text = true }
      )
      :wait()
    eq(
      "workspace: and " .. name .. " has no ws/spec branch left behind",
      vim.trim(left.stdout or ""),
      ""
    )
  end

  -- And it refuses when there IS work.
  local ws2 = workspace.create(m, { name = "spec2", root = project })
  if ws2 then
    local dir = vim.fs.joinpath(ws2.root, "clm")
    local fd = assert(io.open(vim.fs.joinpath(dir, "f.txt"), "w"))
    fd:write "real work\n"
    fd:close()
    vim.system({ "git", "-C", dir, "add", "f.txt" }, { text = true }):wait()
    vim
      .system({ "git", "-C", dir, "-c", "commit.gpgsign=false", "commit", "-qm", "work" }, { text = true })
      :wait()

    local blockers = workspace.unsaved(ws2)
    truthy("workspace: an unpushed commit blocks removal", #blockers > 0, vim.inspect(blockers))
    truthy("workspace: and says why", (blockers[1] or ""):find "unpushed" ~= nil, blockers[1])

    -- A repository holding a DIRECTORY named after its base branch. kora-backend
    -- is one: `main/` sits next to `main`, git calls `HEAD --not --remotes main`
    -- ambiguous and fails, `git()` returns nil, and a nil answer to "is there
    -- unpushed work?" read as NO -- so the one repo where the check could not
    -- run was the one where removal went ahead and took the commits with it.
    vim.fn.mkdir(vim.fs.joinpath(dir, "main"), "p")
    local collide = assert(io.open(vim.fs.joinpath(dir, "main", "x.txt"), "w"))
    collide:write "x\n"
    collide:close()
    vim.system({ "git", "-C", dir, "add", "-A" }, { text = true }):wait()
    vim
      .system(
        { "git", "-C", dir, "-c", "commit.gpgsign=false", "commit", "-qm", "path named main" },
        { text = true }
      )
      :wait()
    local ambiguous = workspace.unsaved(ws2)
    truthy(
      "workspace: a base branch that is also a path still blocks removal",
      #ambiguous > 0 and (ambiguous[#ambiguous]):find "unpushed" ~= nil,
      vim.inspect(ambiguous)
    )

    truthy("workspace: force removes anyway", workspace.remove(ws2, { force = true }))
  end
end

local function test_ws_init()
  -- `raw`, because plain `ws init` now opens the configuration dialog. The
  -- buffer is the escape hatch behind it, and it is the half that carries the
  -- regression: `:Paseo ws init` named the buffer <root>/.ws/workspace.toml
  -- without creating .ws/, so `:w` failed with E212 "Can't open file for
  -- writing: no such file or directory" -- which reads like a permissions
  -- problem rather than a missing parent directory.
  local project = vim.fn.tempname()
  vim.fn.mkdir(vim.fs.joinpath(project, "repo"), "p")
  local repo = vim.fs.joinpath(project, "repo")
  for _, args in ipairs {
    { "init", "-q", "-b", "main", "." },
    { "config", "user.email", "t@example.com" },
    { "config", "user.name", "t" },
  } do
    vim.system(vim.list_extend({ "git", "-C", repo }, args)):wait()
  end
  local fd = assert(io.open(vim.fs.joinpath(repo, "f.txt"), "w"))
  fd:write "x\n"
  fd:close()
  vim.system({ "git", "-C", repo, "add", "-A" }):wait()
  vim.system({ "git", "-C", repo, "-c", "commit.gpgsign=false", "commit", "-qm", "init" }):wait()

  in_dir(project, function()
    vim.cmd "Paseo ws init raw"
    vim.wait(2000, function()
      return vim.api.nvim_buf_get_name(0):find "workspace%.toml" ~= nil
    end, 50)

    local named = vim.api.nvim_buf_get_name(0)
    truthy(
      "ws init raw: the buffer is named after the manifest",
      named:find "%.ws/workspace%.toml" ~= nil,
      named
    )
    truthy(
      "ws init raw: .ws/ exists BEFORE you write",
      vim.uv.fs_stat(vim.fs.dirname(named)) ~= nil
    )

    local ok = pcall(vim.cmd, "silent write")
    truthy("ws init raw: :w succeeds", ok)
    truthy("ws init raw: the manifest is on disk", vim.uv.fs_stat(named) ~= nil)

    -- A second init must reuse the buffer already sitting on that path rather
    -- than failing on a duplicate name.
    truthy("ws init raw: running it twice does not error", pcall(vim.cmd, "Paseo ws init raw"))

    local loaded = require("paseo.workspace.manifest").load(project)
    truthy("ws init raw: what it wrote parses back", loaded ~= nil and loaded.repos.repo ~= nil)
    vim.cmd "silent tabonly"
  end)

  -- The manifest buffer outlives the suite otherwise, and the next time
  -- anything checks timestamps Neovim reports `E211: File ... no longer
  -- available` for a fixture we deleted on purpose.
  for _, buf in ipairs(vim.api.nvim_list_bufs()) do
    if vim.api.nvim_buf_get_name(buf):find(project, 1, true) == 1 then
      pcall(vim.api.nvim_buf_delete, buf, { force = true })
    end
  end
  vim.fn.delete(project, "rf")
end

--- Which project a workspace is nested under, in the editor's view of it.
---
--- The daemon cannot get this right and it is not its fault. A `ws` workspace
--- at `<project>/.workspaces/<name>` is a plain directory holding several
--- worktrees -- it has to be, because a git worktree is per repo -- so asking
--- the daemon to open it registers that directory as its own TOP-LEVEL project
--- named after the workspace. `~/Code/openfin/.workspaces/billing` came back as
--- a project called `billing`, drawn as a sibling of `openfin` rather than as
--- something inside it, which is exactly how the sidebar rendered it.
local function test_workspace_group()
  local workspaces = require "paseo.workspaces"

  eq(
    "ws group: a .workspaces child is grouped under its project",
    workspaces.group { directory = "/home/x/Code/openfin/.workspaces/billing" },
    "openfin"
  )
  eq(
    "ws group: however deep the path goes",
    workspaces.group { directory = "/home/x/Code/grasslabs/kora/.workspaces/nvim-e2e" },
    "kora"
  )
  eq(
    "ws group: an ordinary checkout keeps the daemon's project",
    workspaces.group { directory = "/home/x/Code/paseo.nvim", project = "paseo.nvim" },
    "paseo.nvim"
  )
  eq(
    "ws group: and a worktree under one does too",
    workspaces.group {
      directory = "/home/x/.paseo/worktrees/abc/ui-fixes",
      project = "paseo.nvim",
    },
    "paseo.nvim"
  )
  -- Nothing to go on is an empty string, never nil: it is a sort key and a
  -- table index in the panel.
  eq("ws group: nothing known is empty, not nil", workspaces.group {}, "")

  -- The directory name is configurable, and the shape test has to follow it.
  require("paseo.config").setup { workspaces = { dir = ".units" } }
  eq(
    "ws group: the marker directory follows the config",
    workspaces.group { directory = "/home/x/Code/openfin/.units/billing" },
    "openfin"
  )
  eq(
    "ws group: and `.workspaces` stops being one",
    workspaces.group { directory = "/home/x/Code/openfin/.workspaces/billing", project = "billing" },
    "billing"
  )
  require("paseo.config").setup {}
end

return {
  { "workspace", test_workspace },
  { "workspace group", test_workspace_group },
  { "ws init", test_ws_init },
}
